[ARM] Fix inlining issue in ARM

CrooseGit opened the pull request:
There is an issue on armv7 where a function wont be inlined due to mismatching target features between caller and callee.
The caller has `HasV8Ops` and `FeatureDotProd` and the callee does not, but AFAIK this should not be a problem.
https://godbolt.org/z/f19h3zT66 is an example showing how the call is not inlined on armv7.
The expected asm output would be something like:
```asm
.fnstart
	vsdot.s8	q0, q1, d4[0]
	bx	lr
.Lfunc_end0:

```
Thanks to @Amichaxx  we managed to narrow it down and now can resolve this problem by adding `ARM::FeatureDotProd, ARM::HasV8Ops` to InlineFeaturesAllowed in llvm/lib/Target/ARM/ARMTargetTransformInfo.h, after which the inlining occurs successfully.

Whilst we're at it we have also added some debugging to make it easier to tell why (or why not) a function is being inlined for ARM, and a couple other features that seem to be missing from the list.

This patch was motivated by an issue experienced with rust that was traced back to llvm, and thus was designed to address that.

CrooseGit committed:
Adds HasV8Ops, FeatureDotProd to InlineAllowed

Fixes issue where functions are not inlined when caller has these
features, but callee does not.

adamgemmell commented:
The [original review](https://reviews.llvm.org/D34697#801309) requested an allowlist as missed optimisations are preferable to miscompilations

CrooseGit committed:
Adds debug statements to ARM areInlineCompatible

This makes it easier to see why your function isn't getting inlined for.

CrooseGit committed:
Adds many features to ARM InlineFeaturesAllowed

Removes FeatureD32 and FeatureFP64 from black list in comments as:
- In https://reviews.llvm.org/D34697#805590 D16 and VFPOnlySP were added to this allowlist because they do "the opposite of what you would expect.
- https://github.com/llvm/llvm-project/commit/760df47b778a530e9368a4b8706940ba103d57ba#diff-8165208908f69b3582d556451[…]6c4b474f2bf32c4ac7fec031cf2efd replaces the previous features with the inverse, but incorrectly keeps them in the allow list as the original reasoning no longer applies.

Some subtarget features provide different instructions depending on whether they are set or unset, these features are believed safe as *not* having these features present does not add instructions.

davemgreen reviewed (commented):
Hi - What makes a feature invalid for inlining? If it disables the use of some instructions/registers? It would seem that many more features (including FeatureFP64 and FeatureD32) could be added to the list. Why is HasV8Ops there if HasV8_1aOps isn't? Why is FeatureDotProd, but not FeatureAES?

Can you add a test? Both for inlining a function with dotprod and for inlining _to_ a function with dotprod.

CrooseGit committed:
Adds tests for inlining with dotprod

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.h:
```diff
@@ -64,37 +64,109 @@ class ARMTTIImpl final : public BasicTTIImplBase<ARMTTIImpl> {
+      ARM::FeatureFPRegs16,
+      ARM::FeatureFPRegs,
+      ARM::FeatureAES,
+      ARM::FeatureVFP2_SP,
```
FeatureVFP2_SP is enabled but not all of
FeatureVFP3_D16, FeatureVFP3_D16_SP, FeatureVFP3_SP, 
FeatureFPARMv8_D16, FeatureFPARMv8_D16_SP, FeatureFPARMv8_SP
etc. Should all the _SP be included?

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.h:
```diff
@@ -64,37 +64,109 @@ class ARMTTIImpl final : public BasicTTIImplBase<ARMTTIImpl> {
 
   // Currently the following features are excluded from InlineFeaturesAllowed.
-  // ModeThumb, FeatureNoARM, ModeSoftFloat, FeatureFP64, FeatureD32
+  // ModeThumb, FeatureNoARM, ModeSoftFloat.
```
You added D32 but not F64. Should it be added too?
I would also add 
 - FeatureMatMulInt8, similar to DotProd
 - HasMVEIntegerOps, HasMVEFloatOps, FeatureLOB, FeatureDFB are similar to other features already enabled.
 - FeatureFuseLiterals,FeatureMVEVectorCostFactor1/2/4, FeatureNoPostRASched, FeaturePreferBranchAlign32/64, FeatureSplatVFPToNeon, FeatureUseMIPipeliner, FeatureUseMISched, FeatureUseWideStrideVFP, FeatureAvoidMULS are tuning features so should be safe.

CrooseGit committed:
Responds to feedback

- Changes inlining debug info to only be outputted when inlining fails.
- Adds features to inline allow lists, and sorts list alphabetically
- Formatting
- Typos

smithp35 reviewed (commented):
Checking over the mention of FeatureFP64 and FeatureD32 in the comment excluding features. It looks like these used to be (in the comment) FeatureVFPOnlySP and FeatureD16, which were are "negative" features like FeatureNoArm, so it made sense to exclude them. In https://reviews.llvm.org/D60691 these were replaced with the "positive" features, but it looks like the comment got updated via a search/replace. Now the features are "positive" it makes sense to include in the allow list.

One decision we may want to consider is whether to filter out tuning features from both the caller and callee before comparing. For example if I have a callee with some tuning feature, but a caller without it; then assuming everything else is present and the tuning features of the caller are applied during code-generation, we should be able to inline. If that were desirable it could be done in a follow up patch.

I've not got anything else to add. Happy when davemgreen is happy.

adamgemmell commented:
All features not in the allowlist must match exactly, while for features in the allowlist the callee's features must be a subset of the caller's features.

It's not easy for us to work out which features are additive and which aren't (though I'm sure most are nowadays) so we initially just stuck with features that we knew from Rust.

davemgreen reviewed (approved):
LGTM - thanks
